### ISSUE TITLE
RFC: Customer groups

### ISSUE THREAD

**Issue by maarcingebala** [saleor/saleor#13070]
## Problem

Saleor needs to have a way to organize customers into groups, which could be used for multiple purposes such as:
- granting a specific promotion/discount
- assigning a tax class 
- other use cases provided by apps

---

## 1. General Assumptions

Introduce Customer Groups API similar to the current product collections API. Assumptions:
- Group consists of a name and a collection of assigned customers.
- Group can be enabled/disabled in certain channels.

### Acceptance Criteria

- [ ]
- [ ]

### Mutations

```graphql
type Mutation {

  customerGroupCreate(
    input: CustomerGroupCreateInput!
  ): CustomerGroupCreate @doc(category: "Users")


  customerGroupUpdate(
    id: ID!
    input: CustomerGroupUpdateInput!
  ): CustomerGroupUpdate @doc(category: "Users")


  customerGroupDelete(
    id: ID!
  ): CustomerGroupDelete @doc(category: "Users")


  customerGroupChannelListingUpdate(
    id: ID!
    input: CustomerGroupChannelListingUpdateInput!
  ): CustomerGroupChannelListingUpdate @doc(category: "Users")


  customerGroupAddCustomers(
    id: ID!
    customers: [ID!]!
  ): CustomerGroupAddCustomers @doc(category: "Users")


  customerGroupRemoveCustomers(
    id: ID!
    customers: [ID!]!
  ): CustomerGroupRemoveCustomers @doc(category: "Users")

}
```

### Types

```graphql
type CustomerGroup {
  id: ID!
  name: String!
  slug: String!
  
  createdAt: DateTime!
  updatedAt: DateTime!
  
  customers(
    filter: CustomerGroupFilterInput
    where: CustomerGroupWhereInput
    sortBy: CustomerGroupOrder
    before: String
    after: String
    first: Int
    last: Int
  ): CustomerGroupCountableConnection

  privateMetadata: [MetadataItem!]!
  privateMetafield(key: String!): String
  privateMetafields(keys: [String!]): Metadata
  metadata: [MetadataItem!]!
  metafield(key: String!): String
  metafields(keys: [String!]): Metadata
}
```

## 4. Other
- Webhooks
- External references
- Translations
- Bulk mutations

## 5. To-Do list

```[tasklist]
### Tasks
- [ ] Add a draft title or issue reference here
```

**Comment by korycins** [comment]
Would be good to mention that the scope of these changes should also include new subscription types for the webhooks (this could not be clear). 
Also missing declaration of a new query to fetch the groups, right? It is probably obvious but worth adding to rfc + required permission.

**Comment by maarcingebala** [comment]
Updated the RFC:
- Added missing `customerGroups` and `customerGroup` queries in the schema examples
- Added information about adding new subscription types in the mutation assumptions section.

**Comment by fowczarek** [comment]
1. Should we think about sorting or filtering customer groups? 
2. Should `cleardb` command removes customer groups?

TODO:
- Populate db should create at least one customer group. 

Ideas to feature development:
- Auto-assign user to customer group via email domain.

**Comment by patrys** [comment]
Nitpick: it's called "collections" now so the API draft needs an update.

**Comment by aniav** [comment]
@patrys changed all the naming to collections now.

@fowczarek added the todo and ideas.
Ad 1. I'd go with minimal sorting and filtering for the first implementation. Added filtering by name and sorting by name and creation date.
Ad 2. `cleardb` does clear all users who are not staff so I'd say for the first implementation it should clear all collections too. If we find it causes problems after initial implementation we will improve.

**Comment by timuric** [comment]
I don't see description of  `CustomerCollectionFilterInput` it would be useful to keep the same filter inout as `CustomerFilterInput` https://docs.saleor.io/docs/3.x/api-reference/users/inputs/customer-filter-input

**Comment by timuric** [comment]
Should `me` query and `customer` have property `collections` with slugs? This would be useful for front-end to personalize the experience.

**Comment by aniav** [comment]
@timuric added definitions for filters (we have CustomerFilterInput in one place and new `CustomerCollectionFilterInput` in another). Thanks for spotting that!

Added the collections with slugs to parked for future development to not increase the scope of that RFC and keep the changeset minimal 🙌
